python/GafferOSLUI/OSLObjectUI.py

```diff
@@ -41,6 +42,172 @@
+			w.setReadOnly( readOnly )
+
+	def _updateFromPlug( self ) :
+		pass
+
+
+
+##########################################################################
+# _PrimitiveVariablesFooter
+##########################################################################
+
+class _PrimitiveVariablesFooter( GafferUI.PlugValueWidget ) :
+
+	def __init__( self, plug ) :
+
+		row = GafferUI.ListContainer( GafferUI.ListContainer.Orientation.Horizontal )
+
+		GafferUI.PlugValueWidget.__init__( self, row, plug )
+
+		with row :
+
+				GafferUI.Spacer( imath.V2i( GafferUI.PlugWidget.labelWidth(), 1 ) )
+
+				menuButton = GafferUI.MenuButton(
+					image = "plus.png",
+					hasFrame = False,
+					menu = GafferUI.Menu(
+						Gaffer.WeakMethod( self.__menuDefinition ),
+						title = "Add Input"
+					),
+					toolTip = "Add Input"
+				)
+				menuButton.setEnabled( not Gaffer.MetadataAlgo.readOnly( plug ) )
+
+				GafferUI.Spacer( imath.V2i( 1 ), imath.V2i( 999999, 1 ), parenting = { "expand" : True } )
+
+	def _updateFromPlug( self ) :
+
+		self.setEnabled( self._editable() )
+
+	def __menuDefinition( self ) :
+
+		result = IECore.MenuDefinition()
+
+		# TODO - uv
+		labelsAndConstructors = [
+			( "closure", GafferOSL.ClosurePlug ),
+			( "P", functools.partial( Gaffer.V3fPlug, interpretation = IECore.GeometricData.Interpretation.Point ) ),
+			( "Pref", functools.partial( Gaffer.V3fPlug, interpretation = IECore.GeometricData.Interpretation.Point ) ),
+			( "velocity", functools.partial( Gaffer.V3fPlug, interpretation = IECore.GeometricData.Interpretation.Vector ) ),
+			( "width", functools.partial( Gaffer.FloatPlug ) ),
+			( "Cs", functools.partial( Gaffer.Color3fPlug ) ),
+			( "customInt", Gaffer.IntPlug ),
+			( "customFloat", Gaffer.FloatPlug ),
+			( "customVector", functools.partial( Gaffer.V3fPlug, interpretation = IECore.GeometricData.Interpretation.Vector ) ),
+			( "customNormal", functools.partial( Gaffer.V3fPlug, interpretation = IECore.GeometricData.Interpretation.Normal ) ),
+			( "customPoint", functools.partial( Gaffer.V3fPlug, interpretation = IECore.GeometricData.Interpretation.Point ) ),
+			( "customColor", Gaffer.Color3fPlug ),
+			( "customMatrix", Gaffer.M44fPlug ),
+			( "customString", Gaffer.StringPlug ),
+		]
+
+		for label, constructor in labelsAndConstructors :
+
+			result.append(
+				"/" + label,
+				{
+					"command" : functools.partial( Gaffer.WeakMethod( self.__addPlug ), label, constructor ),
+				}
+			)
+
+		return result
+
+	def __addPlug( self, name, plugConstructor ) :
+
+		if plugConstructor == GafferOSL.ClosurePlug:
+			with Gaffer.UndoScope( self.getPlug().ancestor( Gaffer.ScriptNode ) ) :
+				self.getPlug().addChild( GafferOSL.ClosurePlug( name, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
+
+		else:
+			plug = Gaffer.Plug( "primitiveVariable", 
+				direction = Gaffer.Plug.Direction.In,
+				flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic )
+			plug.addChild( Gaffer.StringPlug(
```

Comment:
I think OSLObject needs to provide a simple API for adding inputs, so we're not left with all clients building their own like this and potentially getting it wrong. The Shuffle node is a good model for this - it provides a nested [ChannelPlug](https://github.com/GafferHQ/gaffer/blob/master/include/GafferImage/Shuffle.h#L60) class which provides everything needed. I suggest we provide a similar `PrimitiveVariablePlug` which is constructed from a value plug, and bundles it with a name plug automatically. I think we should even use this approach for the closure plugs, either by ignoring the name plug or not creating it in the first place. This makes it easier to also support an "enabled" plug if we want to.

Reply:
Should this extra near-duplicate of CompoundData::MemberPlug be shared between OSLImage and OSLObject?  Or is this the same as the UI for it, and I should I should make OSLObject::PrimitiveVariablePlug and then duplicate it as OSLImage::ChannelOrLayerPlug?

Reply:
I'm not sure to be honest. In theory they do all have different constraints on the types that `valuePlug()` can take, and the closure case doesn't use the name. But then again, we don't enforce the "must be able to get data from the plug" constraint for CompoundData::MemberPlug when constructing the graph anyway. What would we call this thing if it were shared? NamedValuePlug? NameAndValuePlug?